[feature request] more direct support for sys_days and year_month_day

Currently `sys_days` are read and written with seconds precision, so to parse / serialize strings like "2018-01-06" one has to jump through onerous hoops. `year_month_day` is only supported in TOML for some reason.

Just to show the problems I was getting with this simple task, I found 1 or 2 compiler bugs:

https://godbolt.org/z/zvWPG87Yn

offending code:
```C++
struct JsonType {
    struct DayEntry {
      year_month_day date;
      int      is_work_day;

      struct glaze {
        static constexpr auto value = glz::array(glz::custom<[](DayEntry& obj, const std::string& str, glz::context& ctx) {
                                                   std::istringstream ss(str);
                                                   ss >> std::chrono::parse(std::string("%F"), obj.date);
                                                   ctx.error = ss.fail() ? glz::error_code::syntax_error : glz::error_code::none;
                                                 },
                                                             nullptr>,
                                                 &DayEntry::is_work_day);
      };
    };
    struct Dates {
      std::vector<DayEntry> data;
    } dailytable;
  };
```

gcc produces ICE on this code, but compiles and works properly if I comment everything related to "JsonType" and only use more simple version.
clang is the clear winner with ":error: member access into incomplete type 'DayEntry'" on exact symbol, but it also errs the same way on the more simple version which gcc compiles.
But then I don't understand how any examples with `struct glaze` inside another struct can work since they kinda provide access to their surrounding type.

MSVC 14.50 is similar to clang but confusing initially as it says I only provided 1 argument to std::chrono::parse leading me to believe the main problem is with that function (which leads to the type being undefined):
```
main.cpp(17): error C2027: use of undefined type 'JsonType::DayEntry'
main.cpp(10): note: see declaration of 'JsonType::DayEntry'
main.cpp(17): error C2672: 'std::chrono::parse': no matching overloaded function found
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\chrono(4735): note: could be 'auto std::chrono::parse(const std::basic_string<_Elem,_Traits,_Alloc> &,_Parsable &,std::basic_string<_Elem,_Traits,_Alloc> &,std::chrono::minutes &)'
main.cpp(17): note: 'auto std::chrono::parse(const std::basic_string<_Elem,_Traits,_Alloc> &,_Parsable &,std::basic_string<_Elem,_Traits,_Alloc> &,std::chrono::minutes &)': expects 4 arguments - 1 provided
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\chrono(4728): note: or       'auto std::chrono::parse(const _CharT *,_Parsable &,std::basic_string<_Elem,_Traits,_Alloc> &,std::chrono::minutes &)'
main.cpp(17): note: 'auto std::chrono::parse(const _CharT *,_Parsable &,std::basic_string<_Elem,_Traits,_Alloc> &,std::chrono::minutes &)': expects 4 arguments - 1 provided
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\chrono(4722): note: or       'auto std::chrono::parse(const std::basic_string<_Elem,_Traits,_Alloc> &,_Parsable &,std::chrono::minutes &)'
main.cpp(17): note: 'auto std::chrono::parse(const std::basic_string<_Elem,_Traits,_Alloc> &,_Parsable &,std::chrono::minutes &)': expects 3 arguments - 1 provided
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\chrono(4716): note: or       'auto std::chrono::parse(const _CharT *,_Parsable &,std::chrono::minutes &)'
main.cpp(17): note: 'auto std::chrono::parse(const _CharT *,_Parsable &,std::chrono::minutes &)': expects 3 arguments - 1 provided
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\chrono(4709): note: or       'auto std::chrono::parse(const std::basic_string<_Elem,_Traits,_Alloc> &,_Parsable &,std::basic_string<_Elem,_Traits,_Alloc> &)'
main.cpp(17): note: 'auto std::chrono::parse(const std::basic_string<_Elem,_Traits,_Alloc> &,_Parsable &,std::basic_string<_Elem,_Traits,_Alloc> &)': expects 3 arguments - 1 provided
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\chrono(4703): note: or       'auto std::chrono::parse(const _CharT *,_Parsable &,std::basic_string<_Elem,_Traits,_Alloc> &)'
main.cpp(17): note: 'auto std::chrono::parse(const _CharT *,_Parsable &,std::basic_string<_Elem,_Traits,_Alloc> &)': expects 3 arguments - 1 provided
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\chrono(4697): note: or       'auto std::chrono::parse(const std::basic_string<_Elem,_Traits,_Alloc> &,_Parsable &)'
main.cpp(17): note: 'auto std::chrono::parse(const std::basic_string<_Elem,_Traits,_Alloc> &,_Parsable &)': expects 2 arguments - 1 provided
C:\Program Files\Microsoft Visual Studio\18\Community\VC\Tools\MSVC\14.50.35717\include\chrono(4692): note: or       'auto std::chrono::parse(const _CharT *,_Parsable &)'
main.cpp(17): note: 'auto std::chrono::parse(const _CharT *,_Parsable &)': expects 2 arguments - 1 provided
```

--- a/docs/chrono.md
+++ b/docs/chrono.md
@@ -68,6 +68,52 @@ glz::read_json(tp, "\"2024-12-13T15:30:45-08:00\"");
 glz::read_json(tp, "\"2024-12-13T15:30:45.123456789Z\"");
 ```
 
+### Date-Only Time Points (`sys_days`)
+
+> **Breaking change:** `std::chrono::sys_days` previously serialized as the full ISO 8601 datetime `"YYYY-MM-DDT00:00:00Z"`. It now serializes as the date-only string `"YYYY-MM-DD"`. The reader accepts both forms, so consumers upgrading Glaze on the read side stay compatible; producers that emit the new shape break older readers that strictly require a time component.
+
+Time points whose duration period is exactly `days` (`std::chrono::sys_days`, i.e. `std::chrono::sys_time<std::chrono::days>`) serialize as date-only strings, since the time-of-day at that precision is always zero:
+
+```cpp
+using namespace std::chrono;
+
+sys_days d = year{2024} / month{6} / day{15};
+std::string json = glz::write_json(d).value();  // "2024-06-15"
+
+sys_days parsed{};
+glz::read_json(parsed, "\"2024-06-15\"");
+```
+
+On read, both the bare date form and a full ISO 8601 datetime are accepted; a full datetime is floored to days precision **after** any timezone offset is applied to UTC:
+
+```cpp
+sys_days parsed{};
+glz::read_json(parsed, "\"2024-06-15T15:30:45Z\"");      // parsed == 2024-06-15
+glz::read_json(parsed, "\"2024-06-14T20:00:00-08:00\""); // parsed == 2024-06-15 (UTC)
+```
+
+Footgun: a sender that thinks "June 14 in California" and emits `"2024-06-14T23:00:00-08:00"` (= `2024-06-15 07:00:00Z`) gets `2024-06-15` back, not `2024-06-14`. The date is the UTC date of the instant, not the local date in the offset. If your protocol means "the local civil date," send a bare `"YYYY-MM-DD"` instead of a datetime with offset.
+
+The reader rejects negative-year strings (`"-0001-01-01"`) and 5+ digit years for symmetry with the writer's `[0000, 9999]` range.
+
+### `std::chrono::year_month_day`
+
+`std::chrono::year_month_day` serializes as a `"YYYY-MM-DD"` JSON string:
+
+```cpp
+using namespace std::chrono;
+
+year_month_day ymd{year{2024}, month{6}, day{15}};
+std::string json = glz::write_json(ymd).value();  // "2024-06-15"
+
+year_month_day parsed{};
+glz::read_json(parsed, json);
+```
+
+Unlike `sys_days`, `year_month_day` accepts only the bare date form on read; embed a `sys_days` field if you also need to accept full datetimes.
+
+Years must lie within RFC 3339's four-digit range `[0000, 9999]` on write; values outside that range fail to serialize with `error_code::constraint_violated` rather than emit wrap-around digits.
+
 ### Steady Clock Time Points
 
 `std::chrono::steady_clock::time_point` serializes as a numeric count (time since epoch), since steady clock's epoch is implementation-defined and not meaningful as a calendar time:
@@ -193,6 +239,8 @@ Output:
 |------|-------------|---------|
 | `std::chrono::duration<Rep, Period>` | Numeric count | `12345` |
 | `std::chrono::system_clock::time_point` | ISO 8601 string | `"2024-12-13T15:30:45Z"` |
+| `std::chrono::sys_days` (`sys_time<days>`) | Date string | `"2024-12-13"` |
+| `std::chrono::year_month_day` | Date string | `"2024-12-13"` |
 | `std::chrono::steady_clock::time_point` | Numeric count | `123456789012345` |
 | `glz::epoch_seconds` | Unix seconds | `1702481400` |
 | `glz::epoch_millis` | Unix milliseconds | `1702481400123` |

--- a/include/glaze/core/chrono.hpp
+++ b/include/glaze/core/chrono.hpp
@@ -128,6 +128,75 @@ namespace glz
 
    namespace chrono_detail
    {
+      // Parse `count` decimal digits starting at s[start]. Returns -1 if any character
+      // is not a digit. Precondition: `count > 0` and `count` digits are readable at
+      // s[start..start+count). `count == 0` returns 0 rather than an error.
+      inline int parse_digits(const char* s, size_t start, size_t count) noexcept
+      {
+         int val = 0;
+         for (size_t i = 0; i < count; ++i) {
+            const char c = s[start + i];
+            if (c < '0' || c > '9') return -1;
+            val = val * 10 + (c - '0');
+         }
+         return val;
+      }
+
+      // Write `val` as exactly N zero-padded decimal digits to b starting at ix, advancing ix.
+      // Caller must ensure b has at least N bytes of capacity at ix.
+      template <size_t N, class B>
+      inline void write_digits(B& b, auto& ix, uint64_t val) noexcept
+      {
+         for (size_t i = N; i > 0; --i) {
+            b[ix + i - 1] = static_cast<char>('0' + val % 10);
+            val /= 10;
+         }
+         ix += N;
+      }
+
+      // Parse exactly "YYYY-MM-DD" (10 chars) into a year_month_day.
+      // On failure, sets ec to parse_error and leaves ymd unchanged.
+      // The 10-char length check implicitly caps the year at 9999, which keeps the
+      // writer's [0000, 9999] range symmetric on read; loosening the size check would
+      // break that symmetry, so it's worth holding fixed.
+      inline void parse_ymd(std::string_view str, std::chrono::year_month_day& ymd, error_code& ec) noexcept
+      {
+         if (str.size() != 10) {
+            ec = error_code::parse_error;
+            return;
+         }
+
+         const char* s = str.data();
+
+         const int yr = parse_digits(s, 0, 4);
+         const int mo = parse_digits(s, 5, 2);
+         const int dy = parse_digits(s, 8, 2);
+
+         if (yr < 0 || mo < 0 || dy < 0 || s[4] != '-' || s[7] != '-') {
+            ec = error_code::parse_error;
+            return;
+         }
+
+         // Fast-fail on obviously out-of-range components before constructing year_month_day.
+         // year_month_day::ok() below catches the remaining cases (e.g. Feb 30, leap years).
+         if (mo < 1 || mo > 12 || dy < 1 || dy > 31) {
+            ec = error_code::parse_error;
+            return;
+         }
+
+         using namespace std::chrono;
+         // yr is in [0, 9999] (4-digit parse, validated non-negative above), which fits
+         // inside std::chrono::year's [-32767, 32767] range, so the int -> year conversion
+         // is in range.
+         const auto candidate = year_month_day{year{yr}, month{static_cast<unsigned>(mo)}, day{static_cast<unsigned>(dy)}};
+         if (!candidate.ok()) {
+            ec = error_code::parse_error;
+            return;
+         }
+
+         ymd = candidate;
+      }
+
       // Parse an RFC 3339 / ISO 8601 date-time string into a system_clock time_point.
       // On failure, sets ec to parse_error and leaves value unchanged.
       template <is_system_time_point TP>
@@ -142,22 +211,12 @@ namespace glz
          const char* s = str.data();
          const auto n = str.size();
 
-         auto parse_digits = [&s](size_t start, size_t count) -> int {
-            int val = 0;
-            for (size_t i = 0; i < count; ++i) {
-               const char c = s[start + i];
-               if (c < '0' || c > '9') return -1;
-               val = val * 10 + (c - '0');
-            }
-            return val;
-         };
-
-         const int yr = parse_digits(0, 4);
-         const int mo = parse_digits(5, 2);
-         const int dy = parse_digits(8, 2);
-         const int hr = parse_digits(11, 2);
-         const int mi = parse_digits(14, 2);
-         const int sc = parse_digits(17, 2);
+         const int yr = parse_digits(s, 0, 4);
+         const int mo = parse_digits(s, 5, 2);
+         const int dy = parse_digits(s, 8, 2);
+         const int hr = parse_digits(s, 11, 2);
+         const int mi = parse_digits(s, 14, 2);
+         const int sc = parse_digits(s, 17, 2);
 
          if (yr < 0 || mo < 0 || dy < 0 || hr < 0 || mi < 0 || sc < 0 || s[4] != '-' || s[7] != '-' || s[10] != 'T' ||
              s[13] != ':' || s[16] != ':') {
@@ -207,7 +266,7 @@ namespace glz
                   ec = error_code::parse_error;
                   return;
                }
-               const int tz_hour = parse_digits(pos, 2);
+               const int tz_hour = parse_digits(s, pos, 2);
                if (tz_hour < 0 || tz_hour > 23) {
                   ec = error_code::parse_error;
                   return;
@@ -220,7 +279,7 @@ namespace glz
                   has_tz_colon = true;
                }
                if (pos + 2 <= n) {
-                  const int m = parse_digits(pos, 2);
+                  const int m = parse_digits(s, pos, 2);
                   if (m < 0 || m > 59) {
                      ec = error_code::parse_error;
                      return;

--- a/include/glaze/json/read.hpp
+++ b/include/glaze/json/read.hpp
@@ -4730,7 +4730,12 @@ namespace glz
       }
    };
 
-   // system_clock::time_point: parse from ISO 8601 string
+   // system_clock::time_point: parse from ISO 8601 string.
+   // Time points whose Duration period is exactly `days` (e.g. std::chrono::sys_days)
+   // also accept the bare "YYYY-MM-DD" date form; full ISO 8601 datetimes are still
+   // accepted in that case and floored to days precision. Coarser periods (weeks,
+   // months, years) intentionally fall through to the full ISO 8601 parser so that
+   // user-supplied dates are not silently snapped to a multi-day boundary.
    template <is_system_time_point T>
       requires(not custom_read<T>)
    struct from<JSON, T>
@@ -4742,7 +4747,36 @@ namespace glz
          from<JSON, std::string_view>::template op<Opts>(str, ctx, it, end);
          if (bool(ctx.error)) [[unlikely]]
             return;
+
+         using Duration = typename std::remove_cvref_t<T>::duration;
+         using Period = typename Duration::period;
+         if constexpr (std::ratio_equal_v<Period, std::ratio<86400>>) {
+            if (str.size() == 10) {
+               std::chrono::year_month_day ymd{};
+               chrono_detail::parse_ymd(str, ymd, ctx.error);
+               if (bool(ctx.error)) [[unlikely]]
+                  return;
+               value = std::chrono::time_point_cast<Duration>(std::chrono::sys_days{ymd});
+               return;
+            }
+         }
          chrono_detail::parse_iso8601(str, value, ctx.error);
+      }
+   };
+
+   // year_month_day: parse from "YYYY-MM-DD" JSON string
+   template <is_year_month_day T>
+      requires(not custom_read<T>)
+   struct from<JSON, T>
+   {
+      template <auto Opts, class It0, class It1>
+      static void op(auto&& value, is_context auto&& ctx, It0&& it, It1 end) noexcept
+      {
+         std::string_view str;
+         from<JSON, std::string_view>::template op<Opts>(str, ctx, it, end);
+         if (bool(ctx.error)) [[unlikely]]
+            return;
+         chrono_detail::parse_ymd(str, value, ctx.error);
       }
    };
 

--- a/include/glaze/json/write.hpp
+++ b/include/glaze/json/write.hpp
@@ -2381,95 +2381,158 @@ namespace glz
       }
    };
 
-   // system_clock::time_point: serialize as ISO 8601 string
-   // Zero-allocation implementation writing directly to buffer
+   // system_clock::time_point: serialize as ISO 8601 string.
+   // Zero-allocation implementation writing directly to buffer.
+   // Time points whose Duration period is exactly `days` (e.g. std::chrono::sys_days)
+   // are written as date-only "YYYY-MM-DD" since the time-of-day is always zero at that
+   // precision and the calendar date is the meaningful payload. Coarser periods (weeks,
+   // months, years) fall through to the full ISO 8601 path; this avoids silently
+   // truncating an arbitrary date to a multi-day boundary on read.
    template <is_system_time_point T>
       requires(not custom_write<T>)
    struct to<JSON, T>
    {
       template <auto Opts, class B>
-      static void op(auto&& value, [[maybe_unused]] is_context auto&& ctx, B&& b, auto& ix) noexcept
+      static void op(auto&& value, is_context auto&& ctx, B&& b, auto& ix) noexcept
       {
          using namespace std::chrono;
          using TP = std::remove_cvref_t<T>;
          using Duration = typename TP::duration;
+         using Period = typename Duration::period;
 
-         // Split into date and time-of-day
+         constexpr bool date_only = std::ratio_equal_v<Period, std::ratio<86400>>;
+
          const auto dp = floor<days>(value);
          const year_month_day ymd{dp};
-         const hh_mm_ss tod{floor<Duration>(value - dp)};
-
-         // Extract components
          const int yr = static_cast<int>(ymd.year());
          const unsigned mo = static_cast<unsigned>(ymd.month());
          const unsigned dy = static_cast<unsigned>(ymd.day());
-         const auto hr = static_cast<unsigned>(tod.hours().count());
-         const auto mi = static_cast<unsigned>(tod.minutes().count());
-         const auto sc = static_cast<unsigned>(tod.seconds().count());
 
-         // Calculate fractional digits based on duration precision
-         constexpr size_t frac_digits = []() constexpr {
-            using Period = typename Duration::period;
-            if constexpr (std::ratio_greater_equal_v<Period, std::ratio<1>>) {
-               return 0; // seconds or coarser
-            }
-            else if constexpr (std::ratio_greater_equal_v<Period, std::milli>) {
-               return 3; // milliseconds
-            }
-            else if constexpr (std::ratio_greater_equal_v<Period, std::micro>) {
-               return 6; // microseconds
-            }
-            else {
-               return 9; // nanoseconds or finer
-            }
-         }();
+         // RFC 3339 requires a 4-digit year in [0000, 9999]. Reject anything else rather
+         // than silently corrupting the output via uint64_t wrap of a negative year.
+         if (yr < 0 || yr > 9999) [[unlikely]] {
+            ctx.error = error_code::constraint_violated;
+            return;
+         }
 
-         // Max size: "YYYY-MM-DDTHH:MM:SS.nnnnnnnnnZ" = 30 + quotes = 32
-         constexpr size_t max_size = 22 + (frac_digits > 0 ? 1 + frac_digits : 0);
+         if constexpr (date_only) {
+            // "YYYY-MM-DD" = 10 chars + optional quotes
+            constexpr size_t max_size = 12;
+            if (!ensure_space(ctx, b, ix + max_size)) [[unlikely]] {
+               return;
+            }
+
+            if constexpr (not check_unquoted(Opts)) {
+               b[ix++] = '"';
+            }
+            chrono_detail::write_digits<4>(b, ix, static_cast<uint64_t>(yr));
+            b[ix++] = '-';
+            chrono_detail::write_digits<2>(b, ix, mo);
+            b[ix++] = '-';
+            chrono_detail::write_digits<2>(b, ix, dy);
+            if constexpr (not check_unquoted(Opts)) {
+               b[ix++] = '"';
+            }
+         }
+         else {
+            const hh_mm_ss tod{floor<Duration>(value - dp)};
+            const auto hr = static_cast<unsigned>(tod.hours().count());
+            const auto mi = static_cast<unsigned>(tod.minutes().count());
+            const auto sc = static_cast<unsigned>(tod.seconds().count());
+
+            // Calculate fractional digits based on duration precision
+            constexpr size_t frac_digits = []() constexpr {
+               if constexpr (std::ratio_greater_equal_v<Period, std::ratio<1>>) {
+                  return 0; // seconds or coarser
+               }
+               else if constexpr (std::ratio_greater_equal_v<Period, std::milli>) {
+                  return 3; // milliseconds
+               }
+               else if constexpr (std::ratio_greater_equal_v<Period, std::micro>) {
+                  return 6; // microseconds
+               }
+               else {
+                  return 9; // nanoseconds or finer
+               }
+            }();
+
+            // Max size: "YYYY-MM-DDTHH:MM:SS.nnnnnnnnnZ" = 30 + quotes = 32
+            constexpr size_t max_size = 22 + (frac_digits > 0 ? 1 + frac_digits : 0);
+            if (!ensure_space(ctx, b, ix + max_size)) [[unlikely]] {
+               return;
+            }
+
+            if constexpr (not check_unquoted(Opts)) {
+               b[ix++] = '"';
+            }
+            chrono_detail::write_digits<4>(b, ix, static_cast<uint64_t>(yr));
+            b[ix++] = '-';
+            chrono_detail::write_digits<2>(b, ix, mo);
+            b[ix++] = '-';
+            chrono_detail::write_digits<2>(b, ix, dy);
+            b[ix++] = 'T';
+            chrono_detail::write_digits<2>(b, ix, hr);
+            b[ix++] = ':';
+            chrono_detail::write_digits<2>(b, ix, mi);
+            b[ix++] = ':';
+            chrono_detail::write_digits<2>(b, ix, sc);
+
+            // Write fractional seconds if duration is finer than seconds
+            if constexpr (frac_digits > 0) {
+               b[ix++] = '.';
+               const auto subsec = tod.subseconds();
+               if constexpr (frac_digits == 3) {
+                  chrono_detail::write_digits<3>(b, ix, static_cast<uint64_t>(duration_cast<milliseconds>(subsec).count()));
+               }
+               else if constexpr (frac_digits == 6) {
+                  chrono_detail::write_digits<6>(b, ix, static_cast<uint64_t>(duration_cast<microseconds>(subsec).count()));
+               }
+               else {
+                  chrono_detail::write_digits<9>(b, ix, static_cast<uint64_t>(duration_cast<nanoseconds>(subsec).count()));
+               }
+            }
+
+            b[ix++] = 'Z';
+            if constexpr (not check_unquoted(Opts)) {
+               b[ix++] = '"';
+            }
+         }
+      }
+   };
+
+   // year_month_day: serialize as "YYYY-MM-DD" JSON string
+   template <is_year_month_day T>
+      requires(not custom_write<T>)
+   struct to<JSON, T>
+   {
+      template <auto Opts, class B>
+      static void op(auto&& value, is_context auto&& ctx, B&& b, auto& ix) noexcept
+      {
+         const int yr = static_cast<int>(value.year());
+         const unsigned mo = static_cast<unsigned>(value.month());
+         const unsigned dy = static_cast<unsigned>(value.day());
+
+         // std::chrono::year is a signed 16-bit-ish range, so users can construct dates
+         // outside [0000, 9999]. Reject those rather than emitting wrap-around digits.
+         if (yr < 0 || yr > 9999) [[unlikely]] {
+            ctx.error = error_code::constraint_violated;
+            return;
+         }
+
+         // "YYYY-MM-DD" = 10 chars + optional quotes
+         constexpr size_t max_size = 12;
          if (!ensure_space(ctx, b, ix + max_size)) [[unlikely]] {
             return;
          }
 
-         // Helper to write N-digit zero-padded number
-         auto write_digits = [&]<size_t N>(uint64_t val) {
-            for (size_t i = N; i > 0; --i) {
-               b[ix + i - 1] = static_cast<char>('0' + val % 10);
-               val /= 10;
-            }
-            ix += N;
-         };
-
          if constexpr (not check_unquoted(Opts)) {
             b[ix++] = '"';
          }
-         write_digits.template operator()<4>(static_cast<uint64_t>(yr));
+         chrono_detail::write_digits<4>(b, ix, static_cast<uint64_t>(yr));
          b[ix++] = '-';
-         write_digits.template operator()<2>(mo);
+         chrono_detail::write_digits<2>(b, ix, mo);
          b[ix++] = '-';
-         write_digits.template operator()<2>(dy);
-         b[ix++] = 'T';
-         write_digits.template operator()<2>(hr);
-         b[ix++] = ':';
-         write_digits.template operator()<2>(mi);
-         b[ix++] = ':';
-         write_digits.template operator()<2>(sc);
-
-         // Write fractional seconds if duration is finer than seconds
-         if constexpr (frac_digits > 0) {
-            b[ix++] = '.';
-            const auto subsec = tod.subseconds();
-            if constexpr (frac_digits == 3) {
-               write_digits.template operator()<3>(static_cast<uint64_t>(duration_cast<milliseconds>(subsec).count()));
-            }
-            else if constexpr (frac_digits == 6) {
-               write_digits.template operator()<6>(static_cast<uint64_t>(duration_cast<microseconds>(subsec).count()));
-            }
-            else {
-               write_digits.template operator()<9>(static_cast<uint64_t>(duration_cast<nanoseconds>(subsec).count()));
-            }
-         }
-
-         b[ix++] = 'Z';
+         chrono_detail::write_digits<2>(b, ix, dy);
          if constexpr (not check_unquoted(Opts)) {
             b[ix++] = '"';
          }

--- a/tests/chrono_test/chrono_test.cpp
+++ b/tests/chrono_test/chrono_test.cpp
@@ -937,4 +937,207 @@ suite chrono_roundtrip_1000_tests = [] {
    };
 };
 
+struct DateRecord
+{
+   std::string name;
+   std::chrono::sys_days start;
+   std::chrono::year_month_day end;
+};
+
+suite chrono_date_only_tests = [] {
+   using namespace std::chrono;
+
+   "sys_days_write_date_only"_test = [] {
+      sys_days d = year{2024} / month{6} / day{15};
+      auto json = glz::write_json(d);
+      expect(json.value() == "\"2024-06-15\"") << json.value();
+   };
+
+   "sys_days_roundtrip"_test = [] {
+      sys_days original = year{2024} / month{6} / day{15};
+      auto json = glz::write_json(original);
+
+      sys_days parsed{};
+      expect(!glz::read_json(parsed, json.value()));
+      expect(parsed == original);
+   };
+
+   "sys_days_read_full_iso8601_is_floored"_test = [] {
+      // Accept full ISO 8601 datetime; the time-of-day is discarded since
+      // sys_days has day precision.
+      sys_days parsed{};
+      expect(!glz::read_json(parsed, "\"2024-06-15T15:30:45Z\""));
+      expect(parsed == sys_days{year{2024} / month{6} / day{15}});
+
+      parsed = sys_days{};
+      expect(!glz::read_json(parsed, "\"2024-06-15T23:59:59.999Z\""));
+      expect(parsed == sys_days{year{2024} / month{6} / day{15}});
+   };
+
+   "sys_days_read_with_timezone_offset"_test = [] {
+      // Late-night UTC-08:00 = next day in UTC; floor to that day.
+      sys_days parsed{};
+      expect(!glz::read_json(parsed, "\"2024-06-14T20:00:00-08:00\""));
+      expect(parsed == sys_days{year{2024} / month{6} / day{15}});
+   };
+
+   "sys_days_invalid_date"_test = [] {
+      sys_days parsed{};
+      // February 30 doesn't exist
+      expect(bool(glz::read_json(parsed, "\"2024-02-30\"")));
+      // Wrong separators
+      expect(bool(glz::read_json(parsed, "\"2024/06/15\"")));
+      // Truncated
+      expect(bool(glz::read_json(parsed, "\"2024-06\"")));
+      // Non-digit
+      expect(bool(glz::read_json(parsed, "\"abcd-06-15\"")));
+   };
+
+   "sys_days_epoch"_test = [] {
+      sys_days d{}; // 1970-01-01
+      auto json = glz::write_json(d);
+      expect(json.value() == "\"1970-01-01\"") << json.value();
+   };
+
+   "sys_days_pre_epoch"_test = [] {
+      sys_days d = year{1969} / month{12} / day{31};
+      auto json = glz::write_json(d);
+      expect(json.value() == "\"1969-12-31\"") << json.value();
+
+      sys_days parsed{};
+      expect(!glz::read_json(parsed, json.value()));
+      expect(parsed == d);
+   };
+
+   "year_month_day_write"_test = [] {
+      year_month_day ymd{year{2024}, month{6}, day{15}};
+      auto json = glz::write_json(ymd);
+      expect(json.value() == "\"2024-06-15\"") << json.value();
+   };
+
+   "year_month_day_roundtrip"_test = [] {
+      year_month_day original{year{2030}, month{1}, day{1}};
+      auto json = glz::write_json(original);
+
+      year_month_day parsed{};
+      expect(!glz::read_json(parsed, json.value()));
+      expect(parsed == original);
+   };
+
+   "year_month_day_invalid"_test = [] {
+      year_month_day parsed{};
+      // year_month_day must be exactly 10 chars; no datetime form allowed.
+      expect(bool(glz::read_json(parsed, "\"2024-06-15T00:00:00Z\"")));
+      expect(bool(glz::read_json(parsed, "\"2024-02-30\"")));
+      expect(bool(glz::read_json(parsed, "\"2024-13-01\"")));
+      expect(bool(glz::read_json(parsed, "\"not-a-date\"")));
+   };
+
+   "year_month_day_zero_padding"_test = [] {
+      year_month_day ymd{year{5}, month{1}, day{2}};
+      auto json = glz::write_json(ymd);
+      expect(json.value() == "\"0005-01-02\"") << json.value();
+
+      year_month_day parsed{};
+      expect(!glz::read_json(parsed, json.value()));
+      expect(parsed == ymd);
+   };
+
+   "struct_with_date_types"_test = [] {
+      DateRecord r{"vacation", year{2024} / month{6} / day{15},
+                   year_month_day{year{2024}, month{6}, day{22}}};
+      auto json = glz::write_json(r);
+      expect(json.value() == R"({"name":"vacation","start":"2024-06-15","end":"2024-06-22"})") << json.value();
+
+      DateRecord parsed{};
+      expect(!glz::read_json(parsed, json.value()));
+      expect(parsed.name == r.name);
+      expect(parsed.start == r.start);
+      expect(parsed.end == r.end);
+   };
+
+   "sys_time_days_alias"_test = [] {
+      // sys_time<days> is the same type as sys_days.
+      sys_time<days> tp = sys_days{year{2024} / month{6} / day{15}};
+      auto json = glz::write_json(tp);
+      expect(json.value() == "\"2024-06-15\"") << json.value();
+   };
+
+   "year_month_day_max_year_9999"_test = [] {
+      // Upper boundary of RFC 3339's four-digit year.
+      year_month_day ymd{year{9999}, month{12}, day{31}};
+      auto json = glz::write_json(ymd);
+      expect(json.value() == "\"9999-12-31\"") << json.value();
+
+      year_month_day parsed{};
+      expect(!glz::read_json(parsed, json.value()));
+      expect(parsed == ymd);
+   };
+
+   "year_month_day_min_year_0000"_test = [] {
+      // Lower boundary of RFC 3339's four-digit year.
+      year_month_day ymd{year{0}, month{1}, day{1}};
+      auto json = glz::write_json(ymd);
+      expect(json.value() == "\"0000-01-01\"") << json.value();
+
+      year_month_day parsed{};
+      expect(!glz::read_json(parsed, json.value()));
+      expect(parsed == ymd);
+   };
+
+   "year_month_day_year_out_of_range_rejected"_test = [] {
+      // Years outside [0000, 9999] cannot be expressed in RFC 3339's 4-digit form;
+      // the writer must surface an error rather than emit wrap-around digits.
+      year_month_day neg{year{-1}, month{1}, day{1}};
+      auto json_neg = glz::write_json(neg);
+      expect(!json_neg.has_value());
+
+      year_month_day big{year{10000}, month{1}, day{1}};
+      auto json_big = glz::write_json(big);
+      expect(!json_big.has_value());
+   };
+
+   "sys_days_year_out_of_range_rejected"_test = [] {
+      // Same guard applies to the date-only sys_days writer path.
+      sys_days big = sys_days{year{10000} / month{1} / day{1}};
+      auto json = glz::write_json(big);
+      expect(!json.has_value());
+   };
+
+   "sys_time_seconds_year_out_of_range_rejected"_test = [] {
+      // And to the full-ISO branch (the year{-1} cast was previously truncated to
+      // bottom-four digits silently).
+      sys_time<seconds> tp = sys_days{year{-1} / month{1} / day{1}};
+      auto json = glz::write_json(tp);
+      expect(!json.has_value());
+   };
+
+   "year_zero_leap_feb_29"_test = [] {
+      // Year 0 (= 1 BCE proleptic Gregorian) is a leap year (divisible by 400).
+      // Pins down that year_month_day::ok() accepts it and the writer emits "0000-02-29".
+      year_month_day ymd{year{0}, month{2}, day{29}};
+      auto json = glz::write_json(ymd);
+      expect(json.value() == "\"0000-02-29\"") << json.value();
+
+      year_month_day parsed{};
+      expect(!glz::read_json(parsed, json.value()));
+      expect(parsed == ymd);
+   };
+
+   "year_month_day_rejects_negative_year_string"_test = [] {
+      // parse_digits treats '-' as non-digit and returns -1, so leading minus
+      // is rejected. Symmetric with the writer's [0000, 9999] guard.
+      year_month_day parsed{};
+      expect(bool(glz::read_json(parsed, "\"-0001-01-01\"")));
+   };
+
+   "sys_days_full_datetime_nanos_floor"_test = [] {
+      // Last instant of 2024-06-15 in UTC, with nanosecond fraction. Flooring
+      // toward epoch must yield 2024-06-15, not round up to 2024-06-16.
+      sys_days parsed{};
+      expect(!glz::read_json(parsed, "\"2024-06-15T23:59:59.999999999+00:00\""));
+      expect(parsed == sys_days{year{2024} / month{6} / day{15}});
+   };
+};
+
 int main() { return 0; }

